virtool/tasks/spawner.py

```diff
@@ -1,100 +1,244 @@
+
+from virtool.analyses.tasks import StoreNuvsFilesTask
+from virtool.hmm.tasks import HMMRefreshTask
+from virtool.indexes.tasks import EnsureIndexFilesTask
+from virtool.jobs.tasks import TimeoutJobsTask
+from virtool.references.tasks import RefreshReferenceReleasesTask, CleanReferencesTask
+from virtool.samples.tasks import CompressSamplesTask, MoveSampleFilesTask
+from virtool.startup import get_scheduler_from_app
+from virtool.subtractions.tasks import (
+    AddSubtractionFilesTask,
+    CheckSubtractionsFASTATask,
+)
 from virtool.tasks.data import TasksData
-from virtool.tasks.task import BaseTask
-from virtool.types import App
-
-logger = logging.getLogger("task_spawner")
-
 
-async def startup_databases_for_spawner(app: App):
+from virtool.types import App
+from virtool.utils import timestamp
+from virtool.config.cls import PeriodicTaskSpawnerConfig
+from virtool.tasks.task import BaseTask
+from virtool.tasks.models import Task as SQLTask
+from virtool.tasks.utils import (
+    startup_databases_for_spawner,
+    startup_datalayer_for_spawner,
+)
+from virtool.uploads.tasks import MigrateFilesTask
+import aiohttp
+import aiohttp.web
+from aiohttp.web import Application
+import aiojobs
+import aiojobs.aiohttp
+
+import virtool.http.accept
+import virtool.http.errors
+from virtool.shutdown import (
+    shutdown_client,
+    shutdown_executors,
+    shutdown_redis,
+    shutdown_scheduler,
+)
+from virtool.startup import (
+    startup_executors,
+    startup_http_client,
+    startup_version,
+)
+from virtool.tasks.api import TasksSpawnerView
+
+logging.basicConfig(level=logging.DEBUG)
+
+logger = logging.getLogger("periodic_task_spawner")
+
+
+@dataclass
+class PeriodicTaskRegistration:
     """
-    Connects to MongoDB, Redis and Postgres concurrently
-
-    :param app: the app object
-
+    A dataclass that holds information about a periodic task registration.
     """
-    config = get_config_from_app(app)
-
-    pg, redis = await asyncio.gather(
-        connect_pg(config.postgres_connection_string),
-        connect(config.redis_connection_string),
-    )
-
-    dispatcher_interface = DispatcherClient(redis)
-    await get_scheduler_from_app(app).spawn(dispatcher_interface.run())
 
-    app.update(
-        {
-            "dispatcher_interface": dispatcher_interface,
-            "pg": pg,
-            "redis": redis,
-        }
+    task: Type[BaseTask]
+    interval: float
+    task_rerun_time: float = None
+    last_triggered: Optional[datetime] = None
+
+
+class TaskSpawnerService:
+    def __init__(self, pg, tasks_datalayer: TasksData):
+        self._registered = []
+        self._pg = pg
+        self._tasks_datalayer = tasks_datalayer
+
+    async def prepare(self, tasks: List[Tuple[Type[BaseTask], float]]):
+        """
+        Prepares tasks to be run.
+
+        :param tasks: list of tasks and their corresponding intervals
+        """
+        await self.register(tasks)
+
+        await self.run()
```

Comment:
This function is as written is more of an `initial_run` function than it is prepare since it runs the main event loop. I would be tempted to drop this function entirely, initialize the tasks with `last_triggered=none` field in `__init__` and then at the start of `run` have it populate the `last_triggered` field.